posthog/queries/session_recordings/session_recording.py

```diff
@@ -94,10 +100,81 @@ def get_metadata(self) -> RecordingMetadata:
             distinct_id=first_snapshot.distinct_id,
         )
+        if include_active_segments:
+            recording_metadata["active_segments_by_window_id"] = self._get_active_segments_by_window_id(all_snapshots)
+
+        return recording_metadata
+
+    def _is_active_event(self, event: SnapshotData) -> bool:
+        # Determines which rr-web events are "active" - meaning user generated
+        # Event type 3 means incremental_update (not a full snapshot, metadata etc)
+        # And the following are the defined source types:
+        # Mutation = 0
+        # MouseMove = 1
+        # MouseInteraction = 2
+        # Scroll = 3
+        # ViewportResize = 4
+        # Input = 5
+        # TouchMove = 6
+        # MediaInteraction = 7
+        # StyleSheetRule = 8
+        # CanvasMutation = 9
+        # Font = 10
+        # Log = 11
+        # Drag = 12
+        # StyleDeclaration = 13
+        return event.get("type") == 3 and event.get("data", {}).get("source") in [1, 2, 3, 4, 5, 6, 7, 12]
+
+    def _get_active_segments(self, snapshots):
+        # Takes a list of snapshots and returns a list of active segments with start and end times
+        snapshot_data_list = [event.snapshot_data for event in list(snapshots)]
+        decompressed_data = paginate_chunk_decompression(self._team.pk, self._session_recording_id, snapshot_data_list)
+        active_recording_segments: List[Dict[str, datetime]] = []
+
+        current_active_segment: Optional[Dict[str, datetime]] = None
+
+        for data in decompressed_data.paginated_list:
+            if self._is_active_event(data):
+                current_timestamp = datetime.fromtimestamp(data.get("timestamp", 0) / 1000, timezone.utc)
+
+                # If the time since the last active event is less than the threshold, continue the existing segment
+                if current_active_segment and (current_timestamp - current_active_segment["end_time"]) <= timedelta(
+                    seconds=ACTIVITY_THRESHOLD_SECONDS
+                ):
+                    current_active_segment["end_time"] = current_timestamp
+
+                # Otherwise, start a new segment
+                else:
+                    if current_active_segment:
+                        active_recording_segments.append(current_active_segment)
+                    current_active_segment = {
+                        "start_time": current_timestamp,
+                        "end_time": current_timestamp,
+                    }
+
+        # Add the last segment if it hasn't already been added
+        if current_active_segment and (
+            len(active_recording_segments) == 0 or active_recording_segments[-1] != current_active_segment
+        ):
+            active_recording_segments.append(current_active_segment)
+
+        return active_recording_segments
+
+    def _get_active_segments_by_window_id(self, all_snapshots):
+        snapshots_by_window_id = defaultdict(list)
```

Comment:
Echoing @mariusandra but this might be a memory concern especially if `all_snapshots` is large. I think we're storing each snapshot three times, once in each function stack. This might not be a problem if we find out this isn't a bottleneck though so I'm excited to try it out in prod!

Reply:
Just replying to this comment, so not a full review... 

If the existing implementation anyway decompressed everything, I don't think it's a big issue to do so here as well. As in, it won't make things substantially worse. However I do fear the day when we get a 4GB recording, and manage to kill the server with it. Chunking the compression (convert to a stream somehow?) or preprocessing (e.g. in celery... or then compute the metadata on the fly during ingestion?) could be a more permanent solution, but I'd get this out first instead of prematurely optimising.

Reply:
Yeah I think computing metadata on ingestion would make a ton of sense. It could line up with the work needed to move recording storage to s3 as we would likely end up with a `recordings` table.

In the meantime, I did a bit of cleanup to make sure we're not passing around the full recordings:
* The local variable now stores only the timestamp of active snapshots instead of the entire recording
* The decompression uses a generator when processing the entire thing.